### PR TITLE
[Merged by Bors] - feat(new sink): Initial `azure_monitor_logs` sink

### PR DESCRIPTION
**Pull request by nazar554**
Solves #1808 
Will add validations for configuration and unit tests

### CODE REVIEW

**Comment by nazar554** [comment]
Right now I assume by default that events have `timestamp` field that is ISO8601. Is there a way to make sure that it's the correct field and it's not using Unix format?

There seems to be an issue when a single event is batched in a JSON array. Azure seems to ignore the timestamp field, maybe because I don't drop milliseconds (although it works fine when there are two or more events)

**Comment by lukesteensen** [comment]
> Right now I assume by default that events have `timestamp` field that is ISO8601. Is there a way to make sure that it's the correct field and it's not using Unix format?

By default, events get a timestamp field with an actual `Timestamp` value, not just a string. But that can often be overwritten with user data that can be a string, so it's hard to guarantee anything. Most sinks will read that field and check the type before using it, which it looks like you're doing.

> There seems to be an issue when a single event is batched in a JSON array. Azure seems to ignore the timestamp field, maybe because I don't drop milliseconds (although it works fine when there are two or more events)

That is odd. I didn't see anything that should behave differently with an array of one event.

**Comment by nazar554** [comment]
@lukesteensen 

> Is there some local version of this service that we could add to our integration tests?

I don't know one, you need to have a Azure Log Analytics service in order to obtain `shared_key` and `customer_id`. However, there is free 5 Gb per month plan, so if you have an Azure subscription then it's possible to test this for fee.

> That is odd. I didn't see anything that should behave differently with an array of one event.

I think it's the behavior of the Azure Log Analytics. [Official documentation](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/data-collector-api) suggests a `YYYY-MM-DDThh:mm:ssZ` date format, but if you actually view some logs on Azure Portal it seems they have milliseconds there. It's possible that docs were not updated.
By default `impl Serialize` does nanosecond precision and this makes Azure drop the timestamp entirely (it uses the ingestion time instead).
 So I added one more commit where I format the date manually and use `chrono::SecondsFormat::Millis`.

> We'll also want to add docs, but that can happen in a followup PR.

I think I can add docs as well, I'll try to use some existing sink as a reference

**Comment by nazar554** [comment]
@jamtur01 I am not sure how to build & test the docs, should I clone and build [`timberio/vector-website`](https://github.com/timberio/vector-website) ?

**Comment by jamtur01** [comment]
@nazar554 I am not 100% sure how to test off a branch - @binarylogic?

**Comment by nazar554** [comment]
Anything else I need to do here? Should I also update `vector.spec.toml`?

**Comment by lukesteensen** [comment]
@nazar554 That file gets updated automatically from the config definitions in `.meta/`, so you should be good there.

Once @binarylogic is able to help out with testing the docs I'll make another pass and see if there are any remaining blockers. Thanks again for opening this! Excited to get it merged.

**Comment by binarylogic** [comment]
@nazar554 the docs look good. As long as `make check-meta` works (which is tested in the CI check workflow), then we're good. No need to do anything with the generated files.

**Comment by lukesteensen** [comment]
@binarylogic it looks like something is up with `check-meta` here. Is there something that needs to be updated in the docs?

File: .meta/sinks/azure_monitor_logs.toml.erb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+[sinks.azure_monitor_logs.options.azure_resource_id]
+type = "string"
+common = true
+examples = ["/subscriptions/11111111-1111-1111-1111-111111111111/resourceGroups/otherResourceGroup/providers/Microsoft.Storage/storageAccounts/examplestorage", "/subscriptions/11111111-1111-1111-1111-111111111111/resourceGroups/examplegroup/providers/Microsoft.SQL/servers/serverName/databases/databaseName"]
+required = false
+description = "[Resource ID](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/data-collector-api#request-headers) of the Azure resource the data should be associated with."
+
+<%= render(
+  "_partials/fields/_tls_connector_options.toml",
+  namespace: "sinks.azure_monitor_logs.options",
```
Comment (jamtur01) [inline .meta/sinks/azure_monitor_logs.toml.erb:64]: This needs:

```
enabled_default: false,
```

With the appropriate `true` or `false` value. That'll fix the failing test.

**Review by jamtur01 — CHANGES_REQUESTED** [review]
One minor issue.

**Review by lukesteensen — APPROVED** [review]
This is looking good to me! If we can get CI green then I think we're ready to merge.

Thanks for your patience with this @nazar554, and apologies for it lingering so long. We really appreciate the contribution!

**Review by jamtur01 — CHANGES_REQUESTED** [review]
Looks like some test failures in the CI run.

**Review by jamtur01 — APPROVED** [review]
LGTM.

**Comment by jamtur01** [comment]
@nazar554 I just merged `master` into the branch because there are some CI changes and I want to make sure the merge goes smooth. Will merge post that.

**Comment by jamtur01** [comment]
bors merge

**Pull request closed without merging**